Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm

```diff
@@ -339,122 +318,117 @@
+            auto *base64String = [cocoaPath substringFromIndex:NSMaxRange(base64Range)];
+            auto data = retainPtr([[NSData alloc] initWithBase64EncodedString:base64String options:0]);
+            return API::Data::createWithoutCopying(data)->span();
         }
 
-        if (auto commaRange = [path rangeOfString:@","]; commaRange.location != NSNotFound) {
-            auto *urlEncodedString = [path substringFromIndex:NSMaxRange(commaRange)];
+        if (auto commaRange = [cocoaPath rangeOfString:@","]; commaRange.location != NSNotFound) {
+            auto *urlEncodedString = [cocoaPath substringFromIndex:NSMaxRange(commaRange)];
             auto *decodedString = [urlEncodedString stringByRemovingPercentEncoding];
-            return [decodedString dataUsingEncoding:NSUTF8StringEncoding];
+            auto data = retainPtr([decodedString dataUsingEncoding:NSUTF8StringEncoding]);
+            return API::Data::createWithoutCopying(data)->span();
         }
 
-        ASSERT([path isEqualToString:@"data:"]);
-        return [NSData data];
+        ASSERT([cocoaPath isEqualToString:@"data:"]);
+        return API::Data::createWithoutCopying([NSData data])->span();
     }
 
     // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
-    if ([path hasPrefix:@"/"])
-        path = [path substringFromIndex:1];
+    if ([cocoaPath hasPrefix:@"/"])
+        cocoaPath = [cocoaPath substringFromIndex:1];
 
-    if (id cachedObject = [m_resources objectForKey:path]) {
-        if (auto *cachedData = dynamic_objc_cast<NSData>(cachedObject))
-            return cachedData;
+    if (auto cachedObject = m_resources.getOptional(path)) {
+        if (auto cachedString = String::fromUTF8(*cachedObject); !cachedString.isEmpty())
+            return cachedString.utf8().span();
 
-        if (auto *cachedString = dynamic_objc_cast<NSString>(cachedObject))
-            return [cachedString dataUsingEncoding:NSUTF8StringEncoding];
+        NSData *cachedCocoaObject = static_cast<NSData *>(API::Data::create(*cachedObject)->wrapper());
 
-        ASSERT(isValidJSONObject(cachedObject, JSONOptions::FragmentsAllowed));
+        if (isValidJSONObject(cachedCocoaObject, JSONOptions::FragmentsAllowed)) {
+            auto *cocoaResult = encodeJSONData(cachedCocoaObject, JSONOptions::FragmentsAllowed);
+            RELEASE_ASSERT(cocoaResult);
 
-        auto *result = encodeJSONData(cachedObject, JSONOptions::FragmentsAllowed);
-        RELEASE_ASSERT(result);
+            // Cache the JSON data, so it can be fetched quicker next time.
+            auto result = API::Data::createWithoutCopying(cocoaResult)->span();
+            m_resources.set(path, result);
 
-        // Cache the JSON data, so it can be fetched quicker next time.
-        [m_resources setObject:result forKey:path];
+            return result;
+        }
 
-        return result;
+        return *cachedObject;
     }
```

Comment:
This is doing a lot of work just to get back to the same data result that is in the cache.
```suggestion
    if (auto cachedObject = m_resources.getOptional(path))
        return *cachedObject;
````